### PR TITLE
added leaving reddit dialog

### PR DESCRIPTION
**Pull request by potatoesAreGod**
Fixes: #596 

Changes:
- Spaced out buttons in navbar by another 10px
- Added a popup, asking the user to press confirm before redirecting to Reddit

### CODE REVIEW

**Comment by sigaloid** [comment]
Maybe the URL should be displayed in the popup to provide more context. Thoughts, @potatoesAreGod @Daniel-Valentine @spikecodes ?

**Comment by Daniel-Valentine** [comment]
This is an excellent idea that is executed very well.

I agree with @sigaloid that we should probably display the target URL in the dialog box to give them context. My only other complaints are:

1. the padding fixes to `nav #link` and `#settings_link` which I think should go into another deliverable, unless they're absolutely needed for the dialog box (which I don't think they are)
2. use of the `admin` color variable for the button colors; as a stop gap this is fine, but we should probably define a new color for these buttons, especially since this bright red does clash with some themes; here is an example with Gold 
Screenshot 2022-11-20 at 17-26-55 The Rust Programming Language
    - alternatively, we could make the style for the dialogue box fixed regardless of the theme, and I think I like this option better
3. we should also allow a user to be able to disable this feature if they decide they don't want to deal with the dialog box

It would also be nice if the box were responsive to the escape key, but that's probably not going to happen without Javascript.

**Comment by potatoesAreGod** [comment]
@sigaloid
> Maybe the URL should be displayed in the popup to provide more context.

excellent point. should be easy enough to implement

@Daniel-Valentine 
> the padding fixes to `nav #link` and `#settings_link` which I think should go into another deliverable, unless they're absolutely needed for the dialog box

sorry, what do you mean by this?

> use of the `admin` color variable for the button colors; as a stop gap this is fine, but we should probably define a new color for 
these buttons, especially since this bright red does clash against background in some themes. _[...]_ alternatively, we could make the style for the dialogue box fixed regardless of the theme, and I think I like this option better

color design isn't my strongest side. suggestions for color combinations are welcome

> we should also allow a user to be able to disable this feature if they decide they don't want to deal with the dialog box

this thought struck me too. however, i have no idea on how to implement it. with some time i can probably work it out, although a couple pointers would be appreciated

**Comment by Daniel-Valentine** [comment]
> > the padding fixes to `nav #link` and `#settings_link` which I think should go into another deliverable, unless they're absolutely needed for the dialog box
> 
> sorry, what do you mean by this?

You have two CSS changes, both padding changes to `nav #link` and `#settings_link`, that appear unrelated to the rest of the code changes in this PR. What I'm saying is that it would be better if you had opened a separate PR for these changes, unless they're necessary for rendering the dialog box correctly.

**Comment by Daniel-Valentine** [comment]
@potatoesAreGod: Thanks for implementing the ability for the user to configure whether or not to see the dialog before proceeding to Reddit. I meant to follow up to the remainder of your post with instructions, but it looks like you've pieced together what to do. I'll make an addition to the [wiki](https://github.com/libreddit/libreddit/wiki) with instructions on how to implement new user settings.

Thanks also for reverting the CSS padding changes I mentioned above. (They're good, but we should deliver them in a separate commit; if we need to revert the changes here to support the display of the redirect warning, we won't lose the padding changes this way.)

I'll give the color for the buttons some thought. Likely, we'll have to define a new variable and have the color specified on a per-theme basis.

File: README.md
Comment (Daniel-Valentine) [inline README.md:202]: I would choose a different name for this option. `LINK_CONFIRMATION` is ambiguous; how about `VISIT_REDDIT_CONFIRMATION`?

File: src/utils.rs
Comment (Daniel-Valentine) [inline src/utils.rs:502]: @sigaloid: FYI This is likely to cause conflict with your incoming config changes. I'll leave it to you to decide whether or not this PR should go in first or if we should wait until your config changes have been merged into master.

**Review by Daniel-Valentine — CHANGES_REQUESTED** [review]

File: templates/settings.html
Comment (Daniel-Valentine) [inline templates/settings.html:83]:
```suggestion
				<label for="link_confirmation">Confirm before visiting content on Reddit</label>
```

**Review by Daniel-Valentine — CHANGES_REQUESTED** [review]

**Comment by Daniel-Valentine** [comment]
Couple of thoughts:

1. We should probably make this the default behavior rather than an opt-in feature. That being said, I agree that the setting description should be given in positive language (the current language "Confirm before visiting content on Reddit" is fine), but this option should be enabled if its value is not specified in the environment.
2. This dialog box also needs to show if a user clicks on the "reddit" link within the post itself and not just the Reddit icon in the navigation bar.

File: templates/base.html
Comment (Daniel-Valentine) [inline templates/base.html:49]:
```suggestion
						<p id="reddit_url">https://www.reddit.com{{ url }}</p>
```

File: static/style.css
Comment (Daniel-Valentine) [inline static/style.css:1465]: It doesn't show here, but `git diff` reports there are unnecessary spaces in what is supposed to be a blank line here.

File: app.json
Comment (Daniel-Valentine) [inline app.json:45]: We might want to change this to `true`.

**Comment by Daniel-Valentine** [comment]
@potatoesAreGod:

I gave it some thought and I'm thinking that the best solution would be a unique style definition for the popup, which would be consistent across all possible styles.



Diff below if you'd like to apply it:

```diff
diff --git a/static/style.css b/static/style.css
index 36d9648..1c4a0c4 100644
--- a/static/style.css
+++ b/static/style.css
@@ -4,6 +4,28 @@
 :root {
 	--nsfw: #ff5c5d;
 	--admin: #ea0027;
+
+	/* Reddit redirect warning constants */
+	--popup-red: #ea0027;
+	--popup-black: #111111;
+	--popup-text: white;
+	--popup-background-1: #0f0f0f;
+	--popup-background-2: #220f0f;
+	--popup-reddit-url: var(--popup-red);
+
+	--popup-background: repeating-linear-gradient(
+		-45deg,
+		var(--popup-background-1),
+		var(--popup-background-1) 50px,
+		var(--popup-background-2) 50px,
+		var(--popup-background-2) 100px
+	);
+
+	--popup-toreddit-background: var(--popup-black);
+	--popup-toreddit-text: var(--popup-red);
+	--popup-goback-background: var(--popup-red);
+	--popup-goback-text: #222;
+	--popup-border: 1px solid var(--popup-red);
 }
 
 @font-face {
@@ -143,7 +165,7 @@ nav #libreddit {
 	height: 100vh;
 	bottom: 0;
 	right: 0;
-	background: #000000aa;
+	backdrop-filter: blur(0.25rem) brightness(15%);
 	visibility: hidden;
 	transition: all 0.1s ease-in-out;
 	z-index: 2;
@@ -159,8 +181,8 @@ nav #libreddit {
 	width: fit-content;
 	height: fit-content;
 	padding: 1rem;
-	background: var(--foreground);
-	border: var(--panel-border);
+	background: var(--popup-background);
+	border: var(--popup-border);
 	border-radius: 5px;
 	transition: all 0.1s ease-in-out;
 }
@@ -173,9 +195,14 @@ nav #libreddit {
 }
 
 .popup-inner h1 {
+	color: var(--popup-text);
 	margin: 1.5rem 1.5rem 1rem;
 }
 
+.popup-inner p {
+	color: var(--popup-text);
+}
+
 .popup-inner a {
 	border-radius: 5px;
 	padding: 2%;
@@ -186,8 +213,8 @@ nav #libreddit {
 }
 
 #goback {
-	background: var(--admin);
-	color: var(--foreground);
+	background: var(--popup-goback-background);
+	color: var(--popup-goback-text);
 }
 
 #goback:not(.selected):hover {
@@ -195,13 +222,14 @@ nav #libreddit {
 }
 
 #toreddit {
-	background: var(--foreground);
-	color: var(--admin);
+	background: var(--popup-toreddit-background);
+	color: var(--popup-toreddit-text);
+	border: 1px solid var(--popup-red);
 }
 
 #toreddit:not(.selected):hover {
-	background: var(--admin);
-	color: var(--foreground);
+	background: var(--popup-toreddit-text);
+	color: var(--popup-toreddit-background);
 }
 
 .popup:target {
@@ -211,7 +239,7 @@ nav #libreddit {
 
 #reddit_url {
 	width: 80%;
-	color: var(--admin);
+	color: var(--popup-reddit-url);
 	font-weight: 600;
 	line-break: anywhere;
 	margin-top: 1rem;
```

Here's how the dialog box looks:

Screenshot 2022-11-26 at 02-37-45 Libreddit Settings

The CSS also adds blur and darkening effects to `.popup` (everything behind the popup window itself). This works in Firefox, although the browser isn't taking a screenshot with the effects actually in effect.

File: static/style.css
Comment (potatoesAreGod) [reply static/style.css:1465]: odd. weird that vscode didn't catch it

**Comment by potatoesAreGod** [comment]
@Daniel-Valentine

great work on the design, looks real awesome!

however, your [browser compatibility](https://developer.mozilla.org/en-US/docs/Web/CSS/backdrop-filter#browser_compatibility) isn't the best. the `backdrop-filter` is not implemented in firefox esr yet and will not be until the [latter half of 2023](https://wiki.mozilla.org/Release_Management/Calendar). instead i implemented a small workaround. additionally, `-webkit-backdrop-filter` is required for safari users.

on further thought, i also agree that the popup should be opt-out.

and i didn't have a thought about the reddit link in the post themselves, will give this one some thought.

thanks for all the feedback and help so far

File: templates/post.html
Comment (Daniel-Valentine) [inline templates/post.html:126]: @Daniel-Valentine: This is going to cause a conflict with #637. Make sure to port these changes into `src/utils.html` in the branch in #637.

File: templates/utils.html
Comment (Daniel-Valentine) [inline templates/utils.html:159]: @Daniel-Valentine: Be warned that this may a merge conflict with the branch in #637.

File: README.md
Comment (Daniel-Valentine) [inline README.md:202]: Since this is now a negative, we should change the name of this variable to `DISABLE_VISIT_REDDIT_CONFIRMATION` to make it clear that this variable being set to `on` now **disables** the warning popup.

File: src/utils.rs
Comment (Daniel-Valentine) [inline src/utils.rs:469]:
```suggestion
	pub disable_visit_reddit_confirmation: String,
```

File: app.json
Comment (Daniel-Valentine) [inline app.json:44]:
```suggestion
    "LIBREDDIT_DEFAULT_DISABLE_VISIT_REDDIT_CONFIRMATION": {
```

File: src/settings.rs
Comment (Daniel-Valentine) [inline src/settings.rs:34]:
```suggestion
	"disable_visit_reddit_confirmation",
```

File: src/utils.rs
Comment (Daniel-Valentine) [inline src/utils.rs:502]:
```suggestion
			disable_visit_reddit_confirmation: setting(&req, "visit_reddit_confirmation"),
```

File: templates/base.html
Comment (Daniel-Valentine) [inline templates/base.html:46]:
```suggestion
				{% if prefs.disable_visit_reddit_confirmation != "on" %}
```

File: templates/settings.html
Comment (Daniel-Valentine) [inline templates/settings.html:83]:
```suggestion
				<label for="disable_visit_reddit_confirmation">Do not confirm before visiting content on Reddit</label>
```

File: templates/base.html
Comment (Daniel-Valentine) [inline templates/base.html:47]: @Daniel-Valentine: When this goes into `templates/utils.html`, make sure to remove the `utils::` namespace in the macro call.

File: templates/settings.html
Comment (Daniel-Valentine) [inline templates/settings.html:82]:
```suggestion
			<div id="disable_visit_reddit_confirmation">
```

(will require concomitant change in `static/style.css`)

**Review by Daniel-Valentine — CHANGES_REQUESTED** [review]

**Comment by Daniel-Valentine** [comment]
This is very close to being finished. I've requested a few changes, but once those are in (along with another below) I think this will be ready to be brought into our code base. We're currently working on synchronizing the Ferrit fork code with Libreddit (#608), which is almost done, but those changes will introduces code that will conflict with that in this branch. To mitigate this, what I'll do is port your code to a separate branch, rebase, and merge and resolve commits along the way. I'll credit you as the commit author so that you receive acknowledgement in both the CREDITS file and the release notes.

Apart from the changes I requested above, you will need to add a the value for `disable_visit_reddit_confirmation` in the restore link; see `templates/settings.html` line 118 (Libreddit master) / 123 (your branch).

```diff
diff --git a/templates/settings.html b/templates/settings.html
index 745c2cc..54b9c7a 100644
--- a/templates/settings.html
+++ b/templates/settings.html
@@ -120,7 +120,7 @@
 
 	<div id="settings_note">
 		<p><b>Note:</b> settings and subscriptions are saved in browser cookies. Clearing your cookies will reset them.</p><br>
-		<p>You can restore your current settings and subscriptions after clearing your cookies using <a href="/settings/restore/?theme={{ prefs.theme }}&front_page={{ prefs.front_page }}&layout={{ prefs.layout }}&wide={{ prefs.wide }}&post_sort={{ prefs.post_sort }}&comment_sort={{ prefs.comment_sort }}&show_nsfw={{ prefs.show_nsfw }}&blur_nsfw={{ prefs.blur_nsfw }}&use_hls={{ prefs.use_hls }}&hide_hls_notification={{ prefs.hide_hls_notification }}&subscriptions={{ prefs.subscriptions.join("%2B") }}&filters={{ prefs.filters.join("%2B") }}">this link</a>.</p>
+		<p>You can restore your current settings and subscriptions after clearing your cookies using <a href="/settings/restore/?theme={{ prefs.theme }}&front_page={{ prefs.front_page }}&layout={{ prefs.layout }}&wide={{ prefs.wide }}&post_sort={{ prefs.post_sort }}&comment_sort={{ prefs.comment_sort }}&show_nsfw={{ prefs.show_nsfw }}&blur_nsfw={{ prefs.blur_nsfw }}&use_hls={{ prefs.use_hls }}&hide_hls_notification={{ prefs.hide_hls_notification }}&disable_visit_reddit_confirmation={{ prefs.disable_visit_reddit_confirmation }}&subscriptions={{ prefs.subscriptions.join("%2B") }}&filters={{ prefs.filters.join("%2B") }}">this link</a>.</p>
 	</div>
 </div>
```

Thanks for your hard work on this, as well as fixing my CSS that would be broken on Firefox ESR.

File: templates/post.html
Comment (Daniel-Valentine) [inline templates/post.html:127]: @Daniel-Valentine: Make sure to remove `utils::` namespace in macro call when this goes into `templates/utils.html`.

Comment (potatoesAreGod) [reply templates/post.html:127]: will keep that in mind

File: templates/utils.html
Comment (potatoesAreGod) [reply templates/utils.html:159]: understood

File: templates/base.html
Comment (Daniel-Valentine) [inline templates/base.html:40]:
```suggestion
				<a id="reddit_link" {% if prefs.disable_visit_reddit_confirmation != "off" %}href="#popup"{% else %}href="https://www.reddit.com" rel="nofollow"{{ url }}"{% endif %}>
```

File: templates/utils.html
Comment (Daniel-Valentine) [reply templates/utils.html:159]: @potatoesAreGod: This was a note to myself -- you don't have to worry about this.

File: templates/base.html
Comment (Daniel-Valentine) [inline templates/base.html:46]:
```suggestion
				{% if prefs.disable_visit_reddit_confirmation != "off" %}
```

File: templates/post.html
Comment (Daniel-Valentine) [inline templates/post.html:125]:
```suggestion
						<a id="reddit_link" {% if prefs.visit_reddit_confirmation != "off" %}href="#popup"{% else %}href="https://reddit.com{{ post.permalink }}" rel="nofollow"{{ url }}"{% endif %}>reddit</a>
```

**Review by Daniel-Valentine — CHANGES_REQUESTED** [review]

**Comment by Daniel-Valentine** [comment]
@potatoesAreGod: I've gone ahead and changed the target branch to our new `leaving-reddit-dialog` branch in this repo. I'm about to merge #637 into master, which would cause this PR into master to be impossible without painstaking merge conflict resolution. I'll take care of that for you by having this go in a separate branch, and I'll work on the merge conflicts there. (You'll still receive full credit for your work here.)

File: templates/base.html
Comment (potatoesAreGod) [reply templates/base.html:40]: i see i've missed a few to rename, will rename the last in a separate commit. but switching from `"on"` to `"off"` will create a logical error where the setting for disabling the popup doesn't work properly

File: templates/base.html
Comment (potatoesAreGod) [reply templates/base.html:46]: see my comment on [your suggestion above](https://github.com/libreddit/libreddit/pull/643/files#r1038503842)

File: templates/post.html
Comment (potatoesAreGod) [reply templates/post.html:125]: see my comment on [your suggestion above](https://github.com/libreddit/libreddit/pull/643/files#r1038503842)

**Comment by Daniel-Valentine** [comment]
Sorry for having left this stale. I'm going to bring this into the branch I reference in [my comment above](https://github.com/libreddit/libreddit/pull/643#issuecomment-1333125127) and will then merge that branch into master. I will have to make sure to review the notes I left here as I execute the merge. Thank you for all of your hard work.

**Comment by Daniel-Valentine** [comment]
@potatoesAreGod: This has now made it into v0.29.0.

**Comment by spikecodes** [comment]
That's just because the PR was linked after it was merged. I'll fix that!

**Pull request merged**
